IslandCounter.java

```diff
@@ -0,0 +1,39 @@
+package assignment4;
+
+
+public class IslandCounter {
+
+    private boolean[][] grid;
+
+    public IslandCounter(boolean[][] grid) {
+        if (grid == null || grid.length == 0) {
+            throw new IllegalArgumentException("Grid cannot be empty.");
+        }
+        this.grid = grid;
+    }
+
+    public int countIslands() {

```

Comment:
The total complexity of this code is O(nlogn). I can modify it (as I said in a comment above) to make it approximately O(n).